hooks.md

```diff
@@ -0,0 +1,90 @@
+
+## CRI-O configuration files for automatically enabling Hooks
+
+The way you enable the hooks above is by editing the OCI Specification to add your hook before running the oci runtime, like runc.  But this is what `CRI-O` and `Kpod create` do for you, so we wanted a way for developers to drop configuration files onto the system, so that their hooks would be able to be plugged in.
+
+One problem with hooks is that the runtime actually stalls execution of the container before running the hooks and stalls completion of the container, until all hooks complete.  This can cause some performance issues.  Also a lot of hooks just check if certain configuration is set and then exit early, without doing anything.  For example the [oci-systemd-hook](https://github.com/projectatomic/oci-systemd-hook) only executes if the command is `init` or `systemd`, otherwise it just exits.  This means if we automatically enable all hooks, every container will have to execute oci-systemd-hook, even if they don't run systemd inside of the container.   Also since there are three stages, prestart, poststart, poststop each hook gets executed three times.
+
+
+
+### Json Definition
+
+We decided to add a json file for hook builders which allows them to tell CRI-O when to run the hook and in which stage.
+CRI-O reads all json files in /usr/share/containers/oci/hooks.d/*.json and /etc/containers/oci/hooks.d and sets up the specified hooks to run.  If the same name is in both directories, the one in /etc/containers/oci/hooks.d takes precedence.
+
+The json configuration looks like this in GO
+```
+// HookParams is the structure returned from read the hooks configuration
+type HookParams struct {
+	Hook          string   `json:"hook"`
+	Stage         []string `json:"stages"`
+	Cmds          []string `json:"cmds"`
+	Annotations   []string `json:"annotations"`
+	HasBindMounts bool     `json:"hasbindmounts"`
+}
+```
+
+| Key    | Description                                                                                                                        | Required/Optional |
+| ------ |----------------------------------------------------------------------------------------------------------------------------------- | -------- |
+| hook   | Path to the hook                                                                                                                   | Required |
+| stages | List of stages to run the hook in: Valid options are `prestart`, `poststart`, `poststop`                                           | Required |
+| cmds   | List of regular expressions to match the command for running the container.  If the command matches a regex, the hook will be run  | Optional |
+| annotations | List of regular expressions to match against the Annotations in the container runtime spec, if an Annotations matches the hook will be run|optional |
```

Comment:
micro nit.  'if an Annotations matches', remove the 's' from Annotations.  Also I don't think it matters, but just in case, it might be good to add a space after 'run' and before the '|'.

Reply:
Fixed